numpy/ma/core.py

```diff
@@ -3233,7 +3217,25 @@ def __setitem__(self, indx, value):
                 _mask[indx] = mval
         elif not self._hardmask:
             # Unshare the mask if necessary to avoid propagation
+            # We want to remove the unshare logic in the future:
             if not self._isfield:
+                if self._sharedmask:
+                    # 2016.01.15 -- v1.11.0
+                    warnings.warn(
+                       "setting an item on a masked array which has a shared "
+                       "mask will also change the original mask array in the "
+                       "future. "
+                       "This affects for example `view = parent[::2]; "
+                       "view[0] = 1` where the mask of `parent` will be "
+                       "changed. "
+                       "It also affects original masks e.g. for "
+                       "`ma_arr = MaskedArray(data, mask); ma_arr[0] = 1` "
+                       "changing the values in `mask`. No copy of the mask "
+                       "will be made, so that changes will propagate both "
+                       "ways.\n"
+                       "To get the future behaviour and silence the warning, "
+                       "set `ma_arr._sharedmask = False`.",
+                       FutureWarning, stacklevel=2)
```

Comment:
Setting the private attribute is ugly, but the only way I can think of. Silencing the warning even more, seems only possible by starting to do reference count checks (i.e. refcount of mask == 1 and it has no base -> no copy necessary)